File: desktop/src/ClipBridge.Core/Security/X25519KeyAgreement.cs
using System.Security.Cryptography;
using Org.BouncyCastle.Crypto.Agreement;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Security;

namespace ClipBridge.Core.Security;

/// <summary>
/// Implementa X25519 com HKDF-SHA256 para derivar uma chave de sessão AES-256.
/// </summary>
public sealed class X25519KeyAgreement : IKeyAgreement, IDisposable
{
    public static readonly int PublicKeySizeBytes = X25519PublicKeyParameters.KeySize;

    private readonly X25519PrivateKeyParameters _privateKey;
    private readonly byte[] _publicKey;
    private bool _disposed;

    public X25519KeyAgreement()
    {
        _privateKey = new X25519PrivateKeyParameters(new SecureRandom());
        _publicKey = _privateKey.GeneratePublicKey().GetEncoded();
    }

    public ReadOnlyMemory<byte> PublicKey => _publicKey;

    public string Fingerprint => Convert.ToHexString(SHA256.HashData(_publicKey).AsSpan(0, 6));

    public byte[] DeriveSessionKey(ReadOnlySpan<byte> peerPublicKey, ReadOnlySpan<byte> salt, ReadOnlySpan<byte> info)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        if (peerPublicKey.Length != PublicKeySizeBytes)
        {
            throw new ArgumentException($"A chave pública X25519 deve ter {PublicKeySizeBytes} bytes.", nameof(peerPublicKey));
        }

        var sharedSecret = new byte[X25519PrivateKeyParameters.SecretSize];
        try
        {
            var agreement = new X25519Agreement();
            agreement.Init(_privateKey);
            agreement.CalculateAgreement(new X25519PublicKeyParameters(peerPublicKey.ToArray(), 0), sharedSecret, 0);

            var hkdf = new HkdfBytesGenerator(new Sha256Digest());
            hkdf.Init(new HkdfParameters(sharedSecret, salt.ToArray(), info.ToArray()));
            var sessionKey = new byte[SessionCipher.KeySizeBytes];
            hkdf.GenerateBytes(sessionKey, 0, sessionKey.Length);
            return sessionKey;
        }
        finally
        {
            CryptographicOperations.ZeroMemory(sharedSecret);
        }
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        CryptographicOperations.ZeroMemory(_privateKey.GetEncoded());
        CryptographicOperations.ZeroMemory(_publicKey);
        _disposed = true;
    }
}

File: mobile/app/src/main/java/com/esousa/clipbridge/security/X25519KeyAgreement.kt
package com.esousa.clipbridge.security

import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import org.bouncycastle.crypto.prng.FixedSecureRandom
import java.security.MessageDigest
import java.security.SecureRandom

/** X25519 com HKDF-SHA256, compatível com a implementação do desktop. */
class X25519KeyAgreement {
    private val privateKey = X25519PrivateKeyParameters(SecureRandom())

    val publicKey: ByteArray = privateKey.generatePublicKey().encoded
    val fingerprint: String
        get() = MessageDigest.getInstance("SHA-256")
            .digest(publicKey)
            .take(6)
            .joinToString("") { "%02X".format(it) }

    fun deriveSessionKey(peerPublicKey: ByteArray, salt: ByteArray, info: ByteArray): ByteArray {
        require(peerPublicKey.size == PUBLIC_KEY_SIZE) { "A chave pública X25519 deve ter $PUBLIC_KEY_SIZE bytes." }

        val sharedSecret = ByteArray(X25519PrivateKeyParameters.SECRET_SIZE)
        try {
            val agreement = X25519Agreement()
            agreement.init(privateKey)
            agreement.calculateAgreement(X25519PublicKeyParameters(peerPublicKey, 0), sharedSecret, 0)

            return ByteArray(SessionCipher.KEY_SIZE).also { sessionKey ->
                HKDFBytesGenerator(SHA256Digest()).apply {
                    init(HKDFParameters(sharedSecret, salt, info))
                    generateBytes(sessionKey, 0, sessionKey.size)
                }
            }
        } finally {
            sharedSecret.fill(0)
        }
    }

    companion object {
        const val PUBLIC_KEY_SIZE = 32
    }
}